### ISSUE TITLE
nominatim indexing problem (nominatim/index.c)

### ISSUE THREAD

**Issue by sporkman** [openstreetmap/Nominatim#965]
During setup on a FreeBSD host, I'm running into an issue that appears to result from the indexing process silently failing (when setup calls the nominatim executable).

Much more information is here, including all the debugging, logs, etc. that brought me to the indexing process as the culprit: https://help.openstreetmap.org/questions/62358/local-nominatim-install-lookups-failing

Short version:

- some important tables seem empty
- from the setup log, we can see that the indexing step does not fail or bail out, but it also does not seem to do anything (lots of "Done 0 in 0..." lines)
- dumping this unindexed db and reloading on a working ubuntu host, running the indexer there, and then bringing the indexed db back to the freebsd host results in a working install
- I also tried a few older versions of the nominatim indexer, same results
- a ktrace of the indexing process shows that it does connect to the db and perform queries

Not sure how much you wanted copied to this ticket, let me know what other information you want.

FreeBSD 11.1
PHP 5.6 and PHP 7.0 (tried with both)
PostgreSQL 9.5
PostGIS 2.4.2
clang/LLVM 4.0.0

**Comment by lonvia** [comment]
For background, what indexing is doing is essentially that:

```
for (int rank = 0; rank <= 30; rank++)
  UPDATE placex SET indexed_status = 0 WHERE rank_search = $rank and indexed_status > 0
```

There is an update trigger installed on the placex table which does the heavy lifting and computes all the additional data needed for geocoding (address, search index etc.) The code for this is in `nominatim/index.c`. There is some additional magic going on there which allows to run the query in multiple threads in parallel. I suspect that either threading doesn't work as expected on FreeBSD or that some variable type is defined differently so that queries are not created correctly.

To start with, enable query logging in postgresql (set `log_min_duration_statement = 0`), run indexing with a single rank (`./nominatim/nominatim -i -d nominatim -P 5432 -t  1 -R 0`) and check in the postgres logs what queries postgres receives.

**Comment by lonvia** [comment]
On second thought, there are likely no objects with search rank 0, so better use rank 4:

```
./nominatim/nominatim -i -d nominatim -P 5432 -t  1 -r 4 -R 4
```

**Comment by sporkman** [comment]
Here you go.  Also just for the record, this is the Luxembourg data.

```
Mar  4 17:12:24 nominatim postgres[27904]: [2-1] LOG:  database system was shut down at 2018-03-04 22:12:23 UTC
Mar  4 17:12:24 nominatim postgres[27904]: [3-1] LOG:  MultiXact member wraparound protections are now enabled
Mar  4 17:12:24 nominatim postgres[27903]: [2-1] LOG:  database system is ready to accept connections
Mar  4 17:12:24 nominatim postgres[27908]: [2-1] LOG:  autovacuum launcher started
Mar  4 17:12:59 nominatim postgres[27916]: [3-1] LOG:  duration: 2.452 ms  parse index_sectors: select geometry_sector,count(*) from placex where rank_search = $1 and indexed_status > 0 group by geometry_sector order by geometry_sector
Mar  4 17:12:59 nominatim postgres[27916]: [4-1] LOG:  duration: 0.337 ms  parse index_sectors_osmline: select geometry_sector,count(*) from location_property_osmline where indexed_status > 0 group by geometry_sector order by geometry_sector
Mar  4 17:12:59 nominatim postgres[27916]: [5-1] LOG:  duration: 0.203 ms  parse index_nosectors: select 0::integer,count(*) from placex where rank_search = $1 and indexed_status > 0
Mar  4 17:12:59 nominatim postgres[27916]: [6-1] LOG:  duration: 0.190 ms  parse index_sector_places: select place_id from placex where rank_search = $1 and geometry_sector = $2 and indexed_status > 0
Mar  4 17:12:59 nominatim postgres[27916]: [7-1] LOG:  duration: 0.068 ms  parse index_nosector_places: select place_id from placex where rank_search = $1 and indexed_status > 0 order by geometry_sector
Mar  4 17:12:59 nominatim postgres[27916]: [8-1] LOG:  duration: 0.085 ms  parse index_sector_places_osmline: select place_id from location_property_osmline where geometry_sector = $1 and indexed_status > 0
Mar  4 17:12:59 nominatim postgres[27916]: [9-1] LOG:  duration: 0.068 ms  parse index_nosector_places_osmline: select place_id from location_property_osmline where indexed_status > 0 order by geometry_sector
Mar  4 17:12:59 nominatim postgres[27917]: [3-1] LOG:  duration: 0.632 ms  parse index_placex: update placex set indexed_status = 0 where place_id = $1
Mar  4 17:12:59 nominatim postgres[27917]: [4-1] LOG:  duration: 0.139 ms  parse index_osmline: update location_property_osmline set indexed_status = 0 where place_id = $1
Mar  4 17:12:59 nominatim postgres[27917]: [5-1] LOG:  duration: 0.100 ms  statement: SET lc_messages TO 'C'
Mar  4 17:12:59 nominatim postgres[27917]: [6-1] LOG:  duration: 0.566 ms  parse placex_details: select placex.osm_type, placex.osm_id, placex.class, placex.type, placex.name, placex.housenumber, placex.country_code, ST_AsText(placex.geometry), placex.admin_level, placex.rank_address, placex.rank_search, placex.parent_place_id, parent.osm_type, parent.osm_id, placex.indexed_status, placex.linked_place_id from placex left outer join placex as parent on (placex.parent_place_id = parent.place_id) where placex.place_id = $1
Mar  4 17:12:59 nominatim postgres[27917]: [7-1] LOG:  duration: 0.682 ms  parse placex_address: select osm_type,osm_id,class,type,distance,cached_rank_address,isaddress from place_addressline join placex on (address_place_id = placex.place_id) where place_addressline.place_id = $1 and address_place_id != place_addressline.place_id order by cached_rank_address asc,osm_type,osm_id
Mar  4 17:12:59 nominatim postgres[27917]: [8-1] LOG:  duration: 0.836 ms  parse placex_names: select (each(name)).key,(each(name)).value from (select name from placex where place_id = $1) as x order by (each(name)).key
Mar  4 17:12:59 nominatim postgres[27917]: [9-1] LOG:  duration: 0.153 ms  parse placex_extratags: select (each(extratags)).key,(each(extratags)).value from (select extratags from placex where place_id = $1) as x order by (each(extratags)).key
Mar  4 17:12:59 nominatim postgres[27916]: [10-1] LOG:  duration: 3.950 ms  bind index_sectors: select geometry_sector,count(*) from placex where rank_search = $1 and indexed_status > 0 group by geometry_sector order by geometry_sector
Mar  4 17:12:59 nominatim postgres[27916]: [10-2] DETAIL:  parameters: $1 = '67108864'
Mar  4 17:12:59 nominatim postgres[27916]: [11-1] LOG:  duration: 0.373 ms  execute index_sectors: select geometry_sector,count(*) from placex where rank_search = $1 and indexed_status > 0 group by geometry_sector order by geometry_sector
Mar  4 17:12:59 nominatim postgres[27916]: [11-2] DETAIL:  parameters: $1 = '67108864'
```

**Comment by lonvia** [comment]
The rank parameter is 0x4000000, so obviously in the wrong byte-order. This likely means that the byteswap functions in `nominatim/postgresql.h` don't work as expected. Could you give https://github.com/openstreetmap/Nominatim/pull/93/ a try? And just for completeness: are you on a machine with big-endian or little-endian?

**Comment by sporkman** [comment]
little-endian (amd64).  I grabbed the diff from above, it applied with a fuzz-factor of 3.  It does look like it's checking for gcc and while I think I've convinced cmake to use gcc6 in place of clang (`cmake -DCMAKE_C_COMPILER=/usr/local/bin/gcc -DCMAKE_CXX_COMPILER=/usr/local/bin/g++ ..`), I'm not going to bet my life that it works.

Regardless, still seeing "67108864" (let me know if you want the entire pgsql log).

```
Mar  5 15:19:11 nominatim postgres[35775]: [10-1] LOG:  duration: 1.538 ms  bind index_sectors: select geometry_sector,count(*) from placex where rank_search = $1 and indexed_status > 0 group by geometry_sector order by geometry_sector
Mar  5 15:19:11 nominatim postgres[35775]: [10-2] DETAIL:  parameters: $1 = '67108864'
Mar  5 15:19:11 nominatim postgres[35775]: [11-1] LOG:  duration: 0.081 ms  execute index_sectors: select geometry_sector,count(*) from placex where rank_search = $1 and indexed_status > 0 group by geometry_sector order by geometry_sector
Mar  5 15:19:11 nominatim postgres[35775]: [11-2] DETAIL:  parameters: $1 = '67108864'
```
For little-endian, do I want byteswap.h included or not?  I'd rather skip all that ifdef stuff and just either include it or not for testing...

**Comment by lonvia** [comment]
Postgres docs says that integers should be in network byte order, so the byteswap is needed for little endian. git blame also tells me that the inclusion of `sys/endian.h` should have fixed the issue on BSD-like systems. So I suggest you look out for that file and figure out if special includes are needed for it. If so, the CmakeLists.txt file might have to be extended to find the right include and add it as a compile flag. If you can't find the file, you need to find out where bswap, __BYTE_ORDER and __LITTLE_ENDIAN are defined in your system.

**Comment by sporkman** [comment]
I think I'm on to something. I see other projects that look for "byteswap.h" which does not exist in FreeBSD.  In sys/endian.h there are a bunch of byteswap functions defined, but they appear to be named differently (see https://www.freebsd.org/cgi/man.cgi?query=byteorder&apropos=0&sektion=9&manpath=FreeBSD+11.0-RELEASE+and+Ports&arch=default&format=html).  

Basically, with all the ifdefs and such removed from postgresql.h, this seems to work:


```
#ifndef POSTGRESQL_H
#define POSTGRESQL_H

#define PG_OID_INT8                     20
#define PG_OID_INT4                     23

#include <sys/endian.h>

#define PGint16(x)      bswap16(x)
#define PGint32(x)      bswap32(x)
#define PGint64(x)      bswap64(x)

const char *build_conninfo(const char *db, const char *username, const char *password, const char *host, const c
har *port);

#endif
```

(note the use of "bswapXX" instead of "bswap_XX")

At first glance this seems to be working, will test further...

**Comment by lonvia** [comment]
Hmm, if I look at the Linux includes, they also suggest that 'bswapXX' is the machine-independent solution. We should probably simply change that but also add a proper detection in the cmake file.

**Comment by sporkman** [comment]
I found this snippet from another project where people were having issues, they also dug up the OS-X/Darwin option:

```
/** Wrappers for OS dependent byte swapping functions */	
#ifdef OS_FREEBSD
#include <sys/endian.h>
#define SCByteSwap16(x) bswap16(x)
#define SCByteSwap32(x) bswap32(x)
#define SCByteSwap64(x) bswap64(x)
#elif OS_DARWIN
#include <libkern/OSByteOrder.h>
#define SCByteSwap16(x) OSSwapInt16(x)
#define SCByteSwap32(x) OSSwapInt32(x)
#define SCByteSwap64(x) OSSwapInt64(x)
#else
#include <byteswap.h>
#define SCByteSwap16(x) bswap_16(x)
#define SCByteSwap32(x) bswap_32(x)
#define SCByteSwap64(x) bswap_64(x)
#endif /* OS_FREEBSD */
```

**Comment by lonvia** [comment]
Can you give #977 a try?

**Comment by sporkman** [comment]
Just a quick follow-up - I'm still on this, should be able to test in the next day or two.

**Comment by sporkman** [comment]
Tested and all seems well here.  Thanks so much for the patch!

**Comment by lonvia** [comment]
Fixed with #977.
